Added stricter range type checks and runtime warnings for ENRICH

craigtaverner opened the pull request:
It has been noted that strange or incorrect error messages are returned if the ENRICH command uses incompatible data types, for example a KEYWORD with value 'foo' using in an int_range match: https://github.com/elastic/elasticsearch/issues/107357

This error is thrown at runtime and contradicts the ES|QL policy of only throwing errors at planning time, while at runtime we should instead set results to null and add a warning. However, we could make the planner stricter and block potentially mismatching types earlier.

However runtime parsing of KEYWORD fields has been a feature of ES|QL ENRICH since it's inception, in particular we even have tests asserting that KEYWORD fields containing parsable IP data can be joined to an ip_range ENRICH index.

In order to not create a backwards compatibility problem, we have compromised with the following:

* Strict range type checking at the planner time for incompatible range types, unless the incoming index field is KEYWORD
* For KEYWORD fields, allow runtime parsing of the fields, but when parsing fails, set the result to null and add a warning

Fixes #107357
Fixes #116799

craigtaverner committed:
Added strict range type checks for ENRICH

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Update docs/changelog/115091.yaml

craigtaverner committed:
Cover range(DATE)->DATETIME case

craigtaverner committed:
Added warnings collection in EnrichQuerySourceOperator

craigtaverner committed:
Fixed failing test

And removed incorrect warnings check from another test. But that test is not providing any warnings, so we have a bug in enrich warnings production.

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into enrich_strict_range_types

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into enrich_strict_range_types

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into enrich_strict_range_types

craigtaverner committed:
Fix warnings propagating back from async request

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

craigtaverner committed:
Improved changelog text, and fixed failing tests

craigtaverner committed:
Refactor ENRICH limitations to more closely match reality

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into enrich_strict_range_types

craigtaverner committed:
Update docs/changelog/115091.yaml

elasticsearchmachine commented:
Hi @craigtaverner, I've updated the changelog YAML for you.

craigtaverner committed:
Enrich has no internal config, so we should not use it for parent source.

This should also cover the many places in test code we use null configurations.

craigtaverner committed:
Fix failing test with new warnings

luigidellaquila commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/EnrichLookupService.java:
```diff
@@ -85,6 +90,35 @@ protected QueryList queryList(TransportRequest request, SearchExecutionContext c
     }
 
+    private static void validateTypes(DataType inputDataType, MappedFieldType fieldType) {
+        if (inputDataType == DataType.UNSUPPORTED) {
+            throw new EsqlIllegalArgumentException("ENRICH cannot match on unsupported input data type");
+        }
+        if (fieldType == null) {
+            throw new EsqlIllegalArgumentException("ENRICH cannot match on non-existent field");
+        }
```
Can these cases really happen? Shouldn't they be prevented at Verification time?
In case, it would be good to have a test for them.

luigidellaquila reviewed (approved):
LGTM, I just left a comment about some runtime validations

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into enrich_strict_range_types

craigtaverner committed:
Make yaml tests more resilient in mixed-cluster and serverless

Some nodes don't have the Source field in the transport request

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into enrich_strict_range_types

craigtaverner committed:
Clean up enrich range-type checks with comments

Added extra tests to verify behaviour of match policies on non-keyword fields. They all behave as keywords (the enrich field is converted to keyword at policy execution time, and the input data is converted to keyword at lookup time).

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into enrich_strict_range_types

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.16  | Commit could not be cherrypicked due to conflicts |
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 115091`

craigtaverner commented:
Backported to 8.17 in https://github.com/elastic/elasticsearch/pull/117130, and we'll not bother with 8.16 at this point
